server/aesdsocket.c

```diff
@@ -0,0 +1,358 @@
+        inet_ntop(their_addr.ss_family, &(((struct sockaddr_in*)&their_addr)->sin_addr), client_ip, sizeof(client_ip));
+        syslog(LOG_DEBUG, "Accepted connection from %s", client_ip);
+
+        /* Dynamically allocate an initial buffer size, later resize as needed*/
+        size_t receive_buf_size = BUF_SIZE;
+        char *buf = (char *)malloc(receive_buf_size);
+        memset(buf, 0, BUF_SIZE);
+        if (buf == NULL)
+        {
+            syslog(LOG_ERR, "Memory allocation failed for receiving buffer");
+            close(new_fd);
+            syslog(LOG_DEBUG, "Closed connection from %s", client_ip);
+            continue;
+        }
+
+        int length;
+        size_t total_received = 0, valid_size = 0;
+        char *end_packet = NULL;
+
+        /* Loop to receive data until a newline is found */
+        do
+        {
+            static int count = 0;
+    
+            /* Resize buffer dynamically if needed from the second count*/
+            if ((total_received + 1 >= receive_buf_size) && (count != 0))
+            {
+                /* Double the size */
+                receive_buf_size *= 2;
+                char *new_buf = (char *)realloc(buf, receive_buf_size);
+                if (new_buf == NULL)
+                {
+                    syslog(LOG_ERR, "Realloc failed for receiving buffer");
+                    free(buf);
+                    close(new_fd);
+                    syslog(LOG_DEBUG, "Closed connection from %s", client_ip);
+                    continue;
+                }
+                memset(new_buf + total_received, 0, receive_buf_size - total_received);
+                buf = new_buf;
+            }
+
+            /* Receive data. ssize_t recv(int sockfd, void buf[.len], size_t len, int flags);*/
+            length = recv(new_fd, buf + total_received, receive_buf_size - total_received - 1, 0);
+            if (length == -1)
+            {
+                syslog(LOG_ERR, "Receive failed");
+                free(buf);
+                close(new_fd);
+                syslog(LOG_DEBUG, "Closed connection from %s", client_ip);
+                continue;
+            }
+
+            total_received += length;
+            end_packet = strchr(buf, '\n');
+            count++;
+
+        } while ((end_packet == NULL) && (length > 0));
+
+    /* Check if we found a newline */
+        if (end_packet != NULL)
+        {
+            /* Append to file only until \n */
+            valid_size = end_packet - buf + 1; 
+            /* Null terminate the string */
+            buf[valid_size] = '\0';
+            fwrite(buf, sizeof(char), valid_size, tmp_file);
+            fflush(tmp_file);
+        }
+        else
+        {
+            syslog(LOG_ERR, "Received without newline\n");
+        }
+            
+        free(buf);
+
+        /* Send back to client */
+        fseek(tmp_file, 0, SEEK_SET);
+
+        size_t send_buf_size = valid_size;
+        char *send_buf = (char *)malloc(send_buf_size);
```

Comment:
@traptibalgi what happens when `send_buf_size` is larger than the space available on the heap?